src/main/java/org/assertj/core/api/ErrorCollector.java

```diff
@@ -46,7 +50,12 @@ public Object intercept(Object obj, Method method, Object[] args, MethodProxy pr
       lastResult.setSuccess(false);
       errors.add(e);
     }
-    return result;
+    if (method != null && !method.getReturnType().isInstance(obj)) {
```

Comment:
>If this approach is desired, the interceptor should generally check if the return type is applicable as with Java 9, methods are no longer invokable if the return type is not compatible what is a threat to type-safety.

Some check like this @raphw?

This is the place where the return type of the method might not be the same as `this` and we need to do something. The `null` is not a good approach I know.